test/tests/router_test.js

```diff
@@ -3327,6 +3327,660 @@ test("intermediateTransitionTo() forces an immediate intermediate transition tha
+    match("/bar").to('bar');
+    match("/baz").to('baz');
+  });
+
+  var fooModelCount = 0, barModelCount = 0, bazModelCount = 0;
+
+  router.updateURL = function(updateUrl) {
+    url = updateUrl;
+    assert.ok(false, "The url was not correctly replaced on initial transition");
+  };
+
+  router.replaceURL = function(replaceURL) {
+    url = replaceURL;
+    assert.ok(true, "The url was replaced correctly on initial transition");
+  };
+
+  var fooHandler = {
+    model: function() {
+      fooModelCount++;
+      router.replaceWith('/bar');
+    }
+  };
+
+  var barHandler = {
+    model: function() {
+      barModelCount++;
+      router.replaceWith('/baz');
+    }
+  };
+
+  var bazHandler = {
+    model: function() {
+      bazModelCount++;
+    }
+  };
+
+  handlers = {
+    foo: fooHandler,
+    bar: barHandler,
+    baz: bazHandler
+  };
+
+  transitionTo(router, '/foo');
+
+  assert.equal(url, '/baz');
+  assert.equal(fooModelCount, 1, 'should call foo model once');
+  assert.equal(barModelCount, 1, 'should call bar model once');
+  assert.equal(bazModelCount, 1, 'should call baz model once');
+});
+
+
+test("Calling replaceWith after initial transition in validation hook should use updateUrl", function(assert) {
+  assert.expect(8);
+
+  map(assert, function(match) {
+    match("/foo").to('foo');
+    match("/bar").to('bar');
+  });
+
+  var fooModelCount = 0, barModelCount = 0;
+
+  router.updateURL = function(updateUrl) {
+    url = updateUrl;
+    assert.ok(true, "updateURL should be used");
+  };
+
+  router.replaceURL = function(replaceURL) {
+    url = replaceURL;
+    assert.ok(false, "replaceURL should not be used");
+  };
+
+  var fooHandler = {
+    model: function() {
+      fooModelCount++;
+      router.replaceWith('/bar');
+    }
+  };
+  var barHandler = {
+    model: function() {
+      barModelCount++;
+    }
+  };
+
+  handlers = {
+    foo: fooHandler,
+    bar: barHandler
+  };
+
+  transitionTo(router, '/bar');
+
+  assert.equal(url, '/bar');
+  assert.equal(barModelCount, 1, 'Bar model should be called once');
+  assert.equal(fooModelCount, 0, 'Foo model should not be called');
+
+  transitionTo(router, '/foo');
+
+  assert.equal(url, '/bar');
+  assert.equal(barModelCount, 2, 'Bar model should be called twice');
+  assert.equal(fooModelCount, 1, 'Foo model should be called once');
+});
+
+test("Calling replaceWith after initial transition in validation hook with multiple redirects should use updateUrl", function(assert) {
+  assert.expect(10);
+
+  map(assert, function(match) {
+    match("/foo").to('foo');
+    match("/bar").to('bar');
+    match("/baz").to('baz');
+  });
+
+  var fooModelCount = 0, barModelCount = 0, bazModelCount = 0;
+
+  router.updateURL = function(updateUrl) {
+    url = updateUrl;
+    assert.ok(true, "updateURL should be used");
+  };
+
+  router.replaceURL = function(replaceURL) {
+    url = replaceURL;
+    assert.ok(false, "replaceURL should not be used");
```

Comment:
I love how you're partially following ye ol' @tomhuda style of "English? double quotes. Code? single quotes." :trollface: 

If you want to map all quote behavior in a specific way that might be worth doing. I didn't review the non-diff so I don't know how consistent it is currently.